doc/lxc.container.conf.sgml.in

```diff
@@ -1503,6 +1503,71 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term>
+            <option>lxc.rootfs.overlay.upperdir</option>
+          </term>
+          <listitem>
+            <para>
+              this option can be used as an alternative to lxc.rootfs.path for specifying
+              the upperdir for the rootfs overlay mount. It must be used in conjunction
+              with lxc.rootfs.overlay.lower.[i].path.
+            </para>
+              a pathname is used for this option, for example
+              <filename>/home/joe/.local/share/lxc/my-container/upper</filename>
+            <para>
+            </para>
+          </listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term>
+            <option>lxc.rootfs.overlay.lower.[i].path</option>
+          </term>
+          <listitem>
+            <para>
+              specify the lower path for the ith lowerdir in the overlay
+            </para>
+            <para>
+              For specifying a directory, a pathname can be used. For a fuse
+              mount, the following syntax is used:
+              <filename>fuse:fuse_binary:[arg_1]:[arg_2]:...:[arg_n]:source</filename>
+              The arguments to the <filename>fuse_binary</filename> are specified with
+              trailing colon-separated values. Before calling the
+              <filename>fuse_binary</filename>, the colons are replaced with spaces.
+              <filename>source</filename> is not a special argument, it just denotes the
+              source of the fuse mount, (e.g. the squash archive for squashfs, the remote
+              filesystem for sshfs).
+              The mount point is auto-generated in
+              <filename>LCX_PATH/LXC_NAME/overlay/lower-[i]</filename> and it is appended
+              to the <filename>fuse_binary</filename> command line.
+              For example, for an unprivileged container "my-container",
+              <filename>fuse:/usr/bin/squashfuse:/home/joe/rootfs.sqhs</filename>
+              will call <filename>/usr/bin/squashfuse /home/joe/rootfs.sqhs
+              /home/joe/.local/share/lxc/my-container/overlay/lower-0</filename>
+            </para>
+          </listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term>
+            <option>lxc.rootfs.overlay.lower.[i].mount_options</option>
```

Comment:
It looks to me like this 'mount_options' is just another way of specifying a portion of .path , is that right?

Ie, I think that the following things are equivalent:

    lxc.rootfs.overlay.lower.0.mount_options = allow_other
    lxc.rootfs.overlay.lower.0.path = fuse:/usr/bin/squashfuse:/images/image0.squashfs

and

    lxc.rootfs.overlay.lower.0.path = fuse:/usr/bin/squashfuse:-o:allow_other:/images/image0.squashfs

Is that right?
If it is, then I'd just ditch the extra-code path that supports '.mount_options'.

Reply:
yes, that's right, @hallyn what do you think?

Reply:
While "mount_type:mount_path" is an established, understood way of specifying lxc.rootfs.path, we do *not* want to over-do the : separation, that is, we do not want to paint ourselves into a corner where some things are not expressible.

In lxc, which is used in background servers in darklit rooms, upgrades ABSOLUTELY must not require container configuration tweaks.

So, given that I don't know what kinds of things usres will want to pass for 'options', I'd rather keep that separate.

Reply:
> While "mount_type:mount_path" is an established, understood way of specifying lxc.rootfs.path, we do _not_ want to over-do the : separation, that is, we do not want to paint ourselves into a corner where some things are not expressible.
> 
> In lxc, which is used in background servers in darklit rooms, upgrades ABSOLUTELY must not require container configuration tweaks.
> 
> So, given that I don't know what kinds of things usres will want to pass for 'options', I'd rather keep that separate.

But what if a token in my path requires a ':'?  Why is an "option" (which is then forced to be prefixed with a '-o') any more likely to require a ':' than another element of the command line array to a program?

I think there is some lxc way to "encode" null terminated arrays into a configuration value, isn't there?  Why not use that?

maybe its just repeating the configuration token?

    lxc.rootfs.overlay.lower.0.path = fuse:/usr/bin/squashfuse
    lxc.rootfs.overlay.lower.0.path = images/image0.squashfs
    lxc.rootfs.overlay.lower.0.path = /path/to/mount/point

Reply:
I don't think repeating the configuration token is currently supported in lxc.

Reply:
> > While "mount_type:mount_path" is an established, understood way of specifying lxc.rootfs.path, we do _not_ want to over-do the : separation, that is, we do not want to paint ourselves into a corner where some things are not expressible
.
> > In lxc, which is used in background servers in darklit rooms, upgrades ABSOLUTELY must not require container configuration tweaks.
> > So, given that I don't know what kinds of things usres will want to pass for 'options', I'd rather keep that separate.
>
> But what if a token in my path requires a ':'?

Then you can prefix with "dir:".

>  Why is an "option" (which is then forced to be prefixed with a '-o') any more likely to require a ':' than another element of the command line array to a program?

Because mount options *can* be horribly complicated and baroque strings.

> I think there is some lxc way to "encode" null terminated arrays into a configuration value, isn't there? Why not use that?

Because in general lxc configuration files should still be parsable by
(ancient) scripts.

> maybe its just repeating the configuration token?
>
> ```
> lxc.rootfs.overlay.lower.0.path = fuse:/usr/bin/squashfuse
> lxc.rootfs.overlay.lower.0.path = images/image0.squashfs
> lxc.rootfs.overlay.lower.0.path = /path/to/mount/point
> ```

That's the worst of both worlds, so I'm saying no right now.

I'm ok with some of the other things we've discussed, but using what
Ariel has now and just adding dir: and loop:, we can support:

lxc.rootfs.overlay.lower.0.path = dir:/home/serge/.local/share/lxc/my:first:disk
lxc.rootfs.overlay.lower.0.overlay = dir:/home/serge/local/share/lxc/overlay

and

lxc.rootfs.overlay.lower.0.path = loop:/var/lib/lxc/c1/r.squashfs
lxc.rootfs.overlay.lower.0.overlay = /var/lib/lxc/c1/rootfs
lxc.rootfs.overlay.lower.0.options = allow_other

Now, what I don't see is how we can support the use of veritysetup.  But
then I've just been using higher level tools like atomfs to do that for
me.

So what would an atomfs config look like?

lxc.rootfs.overlay.lower.0.path = mount-helper:/usr/bin/atomfs:/var/lib/lxc/c1/rfs.squashfs
lxc.rootfs.overlay.lower.0.overlay = dir:/var/lib/lxc/c1/overlay
lxc.rootfs.overlay.lower.0.options = allow_other,enforce-dmverity

?

Reply:
> "you really need to just solve the problem."

Yes, I'm only trying to nail down exactly what problems we do need to solve.  And yes, I'm opinionated about that.  I appreciate your input.